Sources/Workspace/Workspace.swift

```diff
@@ -1422,84 +1413,80 @@ extension Workspace {
 
     /// Computes if dependency resolution is required based on input constraints and pins.
     ///
-    /// - Returns: A tuple with two elements.
-    ///       resolve: If resolution is required.
-    ///       validPins: The pins which are still valid.
+    /// - Returns: Returns whether dependency resolution is required.
     // @testable internal
     func isResolutionRequired(
         root: PackageGraphRoot,
-        dependencies: [RepositoryPackageConstraint],
-        pinsStore: PinsStore
-    ) -> (resolve: Bool, validPins: [RepositoryPackageConstraint]) {
-
-        // Create pinned constraints.
-        let pinConstraints = pinsStore.createConstraints()
-
-        // Create a constraint set to check constraints are mergable.
-        var constraintSet = PackageContainerConstraintSet()
-
-        // The input dependencies should be mergable, otherwise we have bigger problems.
-        for constraint in dependencies {
-            if let mergedSet = constraintSet.merging(constraint) {
-                constraintSet = mergedSet
-            } else {
-                return (true, pinConstraints)
-            }
-        }
+        dependencyManifests: DependencyManifests,
+        pinsStore: PinsStore,
+        extraConstraints: [RepositoryPackageConstraint] = []
+    ) -> Bool {
+        let constraints =
+            root.constraints(config: config) +
+            // Include constraints from the manifests in the graph root.
+            root.manifests.flatMap({ $0.dependencyConstraints(config: config) }) +
+            dependencyManifests.dependencyConstraints() +
+            extraConstraints
+
+        let diagnostics = DiagnosticsEngine()
+        let localProvider = LocalContainerProvider(
+             root: root,
+             currentManifests: dependencyManifests,
+             config: config,
+             diagnostics: diagnostics
+        )
 
-        // Merge all the pin constraints.
-        for pin in pinConstraints {
-            if let mergedSet = constraintSet.merging(pin) {
-                constraintSet = mergedSet
-            }
-        }
+        let resolver = PubgrubDependencyResolver(localProvider)
+        let result = resolver.solve(dependencies: constraints, pinsStore: pinsStore)
 
-        // Compute the valid pins, i.e., the pins which are still valid in the
-        // final merged set.
-        let validPins = pinConstraints.filter({
-            $0.requirement == constraintSet[$0.identifier]
-        })
+        guard !diagnostics.hasErrors else { return true }
+        guard case .success(let bindings) = result else { return true }
 
         // Otherwise, check checkouts and pins.
-        for constraint in constraintSet {
-            let url = constraint.key.path
-            let dependency = managedDependencies[forURL: url]
+        for (container, boundVersion) in bindings {
```

Comment:
I think we can kill this loop now since we've already checked everything that was needed using the resolver.

Reply:
Actually, we still need this. The `LocalPackageContainer`-backed resolver can still return a successful resolution with different bound versions: for example, when changing a dependency from a version to a branch or revision requirement, the resolver will always prefer the branch or revision requirement and will resolve successfully, without trying to checkout the branch or revision.

Reply:
Although I admit that I did find it weird that the resolver would resolve a dependency to a branch without checking it out to see if it had different dependencies.

Reply:
Btw, if you are curious, you can see this in action by commenting out the loop and you will see tests `WorkspaceTests.testRevisionVersionSwitch` and `WorkspaceTests.testLocalVersionSwitch` fail. But those tests pass dependencies as package root dependencies, not as dependencies of the root manifest.

Reply:
Hmm, that seems wrong. We should throw from `getUnversionedDependencies` if it's accessed for non-versioned dependencies.

Reply:
That's correct.